x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/retention/AbstractExpiredJobDataRemover.java

```diff
@@ -68,19 +68,29 @@ private void removeData(WrappedBatchedJobsIterator jobIterator, ActionListener<B
             removeData(jobIterator, listener, isTimedOutSupplier);
             return;
         }
-        long cutoffEpochMs = calcCutoffEpochMs(retentionDays);
-        removeDataBefore(job, cutoffEpochMs,
-            ActionListener.wrap(response -> removeData(jobIterator, listener, isTimedOutSupplier), listener::onFailure));
+
+        calcCutoffEpochMs(job.getId(), retentionDays, ActionListener.wrap(
+                cutoffEpochMs -> {
+                    if (cutoffEpochMs == null) {
+                        removeData(jobIterator, listener, isTimedOutSupplier);
+                    } else {
+                        removeDataBefore(job, cutoffEpochMs, ActionListener.wrap(
+                                response -> removeData(jobIterator, listener, isTimedOutSupplier),
+                                listener::onFailure));
+                    }
+                },
+                listener::onFailure
+        ));
     }
 
     private WrappedBatchedJobsIterator newJobIterator() {
         BatchedJobsIterator jobsIterator = new BatchedJobsIterator(client, AnomalyDetectorsIndex.configIndexName());
         return new WrappedBatchedJobsIterator(jobsIterator);
     }
 
-    private long calcCutoffEpochMs(long retentionDays) {
+    void calcCutoffEpochMs(String jobId, long retentionDays, ActionListener<Long> listener) {
```

Comment:
The other methods that extending classes are expected to override are `protected`, even though the actual classes that do extend this class are all in the same package.  But then this class's constructor is package private so it would be impossible to have a derived class in another package despite the abstract methods being set up for that.  I think for consistency they should all be the same - either `protected` or package private.  Certainly with a default implementation here I think `protected` makes it clearer that we expect derived classes to modify it rather than it's just been made accessible for testing.  But then if you agree with my other suggestion and make this `abstract` then that also makes that clear.

Reply:
The method is package-private for testing otherwise it is very difficult to test and can only be done indirectly. `abstract` makes sense.

Reply:
But `protected` would also allow it to be tested directly.  Basically I think all the `abstract` methods and the constructor should have the same accessibility, whether that be `protected` or package private.  So either change the ones that are currently `protected` to be package private or change this one plus the constructor to be `protected`.

Reply:
Given that the base class is package private, only classes in the same package can implement the abstract method whether the method is protected or package private. The only difference is I could derive a new class from one of the package's public non-abstract classes and reimplement `calcCutoffEpochMs` in a different package if it was protected but not if package private. In practice this isn't a concern so I've gone for the principle of least visibility and made the abstract methods package private.